### PR TITLE
[WIP] Fixes #25811 - Add Dual List component to "Create Architectures" page

### PR DESCRIPTION
**Pull request by Ron-Lavi**
Introduce the DualList component in Foreman

Selection_036

### CODE REVIEW

File: webpack/assets/javascripts/react_app/redux/reducers/index.js [hunk 1]
```diff
@@ -8,6 +8,7 @@ import toasts from './toasts';
 import { reducers as passwordStrengthReducers } from '../../components/PasswordStrength';
 import { reducers as breadcrumbBarReducers } from '../../components/BreadcrumbBar';
 import { reducers as searchBarReducers } from '../../components/SearchBar';
+import { reducers as dualListReducers } from '../../components/DualList';
```
Comment (glekner) [inline webpack/assets/javascripts/react_app/redux/reducers/index.js:11]: Am I missing something ? I dont see any reducers in `index.js`

Comment (Ron-Lavi) [reply webpack/assets/javascripts/react_app/redux/reducers/index.js:11]: they will be added later, still a very very early WIP.

Comment (glekner) [reply webpack/assets/javascripts/react_app/redux/reducers/index.js:11]: didn't see WIP 💃

**Comment by waldenraines** [comment]
Will this make https://projects.theforeman.org/issues/21952 a duplicate of [#25811](https://projects.theforeman.org/issues/25811)?

**Comment by Ron-Lavi** [comment]
@waldenraines I am assuming it won't be needed anymore, 
as we're going to replace the whole implementation
and there won't be any need of the jQuery select library,
Are there more implementations of dual lists that I should be aware of? Thanks

**Comment by ohadlevy** [comment]
@Laviro what is the status of this PR? thanks.

**Comment by Ron-Lavi** [comment]
@ohadlevy sorry, i had to work on some other stuff but soon I will be back to work on it :)

File: webpack/assets/javascripts/react_app/redux/reducers/index.js [hunk 2]
```diff
@@ -8,6 +8,7 @@ import toasts from './toasts';
 import { reducers as passwordStrengthReducers } from '../../components/PasswordStrength';
 import { reducers as breadcrumbBarReducers } from '../../components/BreadcrumbBar';
 import { reducers as autoCompleteReducers } from '../../components/AutoComplete';
+import { reducers as dualListReducers } from '../../components/DualList';
```
Comment (ezr-ondrej) [inline webpack/assets/javascripts/react_app/redux/reducers/index.js:11]: early thought, just something I have been thinking for a while: with redux, what is going to happen, when we got two components of the same kind on the page? :thinking: Will they not start to fight over the same namescpace in redux state tree?

Comment (Ron-Lavi) [reply webpack/assets/javascripts/react_app/redux/reducers/index.js:11]: great thought!
We could take advantage of an ID mechanism, which will be the namespace for the reducer.
I open a PR to [distinguish between autocumpletes](https://github.com/theforeman/foreman/pull/6501) which we can implement here too !

**Review by ohadlevy — COMMENTED** [review]
is this used anywhere? I'm also not seeing the form wrapper? or is this just plain infrastructure? in that case I'm also missing storybook?

why did you decide to break it down?

**Comment by Ron-Lavi** [comment]
@ohadlevy replacing the `multiple_selects` implementation everywhere means this PR would get too big and not sure if it would ever get merge..

I prefer to break it into smaller steps, and later to replace individual implementations to insert this component smoothly. another thought is to let other consumers who want to use this component the chance, instead of blocking them. 

I will add the story book soon :+1:

**Comment by Ron-Lavi** [comment]
Added Australian Dual List storybook :)

Selection_036

@theforeman/ui-ux  ready for review

File: app/views/common/_dual_list.html.erb [hunk 3]
```diff
@@ -0,0 +1,5 @@
+<div id=<%=id%> ></div>
+<%= mount_react_component('ComponentWrapper', "##{id}", {
```
Comment (amirfefer) [inline app/views/common/_dual_list.html.erb]: `ComponentWrapper` is not needed any more, use `flattenData` flag instead

File: webpack/assets/javascripts/react_app/components/DualList/DualList.fixtures.js [hunk 4]
```diff
@@ -0,0 +1,15 @@
+export const id = 'Koala';
```
Comment (amirfefer) [inline webpack/assets/javascripts/react_app/components/DualList/DualList.fixtures.js]: is `export` needed?

File: webpack/assets/javascripts/react_app/components/DualList/DualList.js [hunk 5]
```diff
@@ -0,0 +1,46 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import { DualListControlled as PFDualList } from 'patternfly-react';
```
Comment (amirfefer) [inline webpack/assets/javascripts/react_app/components/DualList/DualList.js]: nit: can you name it `PfDualList` instead

File: webpack/assets/javascripts/react_app/components/DualList/DualList.js [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+
+  return (
+    <PFDualList
+      allowHiddenInputs
+      left={{
+        items: unselectedlist,
+      }}
+      right={{
+        items: selectedList,
+        inputProps: { name: `filter[${inputName}][]` },
```
Comment (amirfefer) [inline webpack/assets/javascripts/react_app/components/DualList/DualList.js]: does `filter` prefix needed for every dual list ?

File: app/views/common/_dual_list.html.erb [hunk 7]
```diff
@@ -0,0 +1,5 @@
+<div id=<%=id%> ></div>
```
Comment (amirfefer) [inline app/views/common/_dual_list.html.erb]: does it need to be partial ? what about helper function?

**Review by amirfefer — CHANGES_REQUESTED** [review]
Thanks @Laviro ! Great work :+1: 

I think consuming this component in this PR is necessary (at least in one page), because the infrastructure might be changed during consumption

**Comment by ohadlevy** [comment]
@Laviro whats the status of this pr? please rebase and @amirfefer @sharvit can you ack/nack? thanks!

**Comment by Ron-Lavi** [comment]
@ohadlevy, I'm going to replace architecture's `create new architecture` dual list by it:

Selection_042

**Comment by Ron-Lavi** [comment]
closing this in favor of #6724

**Pull request closed without merging**